lib/eventHubReceiver.ts

```diff
@@ -0,0 +1,382 @@
+      await defaultLock.acquire(this._context.cbsSession.cbsLock, () => { return this._context.cbsSession.init(this._context.connection); });
+      const tokenObject = await this._context.tokenProvider.getToken(this.audience);
+      debug(`[${this._context.connectionId}] EH Receiver: calling negotiateClaim for audience "${this.audience}"`);
+      // Negotitate the CBS claim.
+      await this._context.cbsSession.negotiateClaim(this.audience, this._context.connection, tokenObject);
+      if (!this._session && !this._receiver) {
+        let receiverError: any;
+        let rcvrOptions: rheaPromise.ReceiverOptions = {
+          autoaccept: true,
+          source: {
+            address: this.address
+          },
+          credit_window: this.prefetchCount,
+        };
+        if (this.epoch !== undefined && this.epoch !== null) {
+          if (!rcvrOptions.properties) rcvrOptions.properties = {};
+          rcvrOptions.properties[Constants.attachEpoch] = rhea.types.wrap_long(this.epoch);
+        }
+        if (this.identifier) {
+          if (!rcvrOptions.properties) rcvrOptions.properties = {};
+          rcvrOptions.properties[Constants.receiverIdentifierName] = this.identifier;
+        }
+        if (this.receiverRuntimeMetricEnabled) {
+          rcvrOptions.desired_capabilities = Constants.enableReceiverRuntimeMetricName;
+        }
+        if (this.options && this.options.eventPosition) {
+          // Set filter on the receiver if event position is specified.
+          let filterClause = this.options.eventPosition.getExpression();
+          if (filterClause) {
+            rcvrOptions.source.filter = {
+              "apache.org:selector-filter:string": rhea.types.wrap_described(filterClause, 0x468C00000004)
+            };
+          }
+        }
+        this._session = await rheaPromise.createSession(this._context.connection);
+        const handleReceiverError = (context: rheaPromise.Context) => {
+          receiverError = errors.translate(context.receiver.error);
+          debug(`An error occurred while creating the receiver "${this.name}" : `, receiverError);
+        };
+        this._session.on(Constants.receiverError, handleReceiverError);
+        debug("Trying to create a receiver...");
+        this._receiver = await rheaPromise.createReceiver(this._session, rcvrOptions);
+        debug("Promise to create the receiver resolved. Created receiver with name: ", this.name);
+
+        this.name = this._receiver.name;
+        if (receiverError) {
+          // There are cases where the EH service sends an attach frame, which causes rhea to emit receiver_open event
+          // thus resolving the promise to create a receiver and moments later the service sends back a detach frame
+          // indicating that there was some error. Hence we check for receiverError, even after the promise has resolved.
+          debug("throwing the receiverError, ", receiverError);
```

Comment:
@grs @princjef  @pierreca - Based on the scenario explanation in the code comment, I chose to handle it like this. Do you think this can be handled in a better way?